File: static/js/ChessPieces/Rook.js
class Rook extends ChessPiece{

	constructor(){
		super();
	}

	movablePlaces(x, y){
		return this.shrinkContinuosArray(this.getRookMovablePlaces(x, y));
	}

	getRookMovablePlaces(x, y){
		var matrix = this.live_chessboard_matrix;
		var placeIds = [];
		var left = [];
		var right = [];
		var up = [];
		var down = [];
		var leftCounter = 0;
		var upCounter = 0;

		if(x>=0&&x<=7&&y>=0&&y<=7){

			//Going Left
			for(var i=7; i>=0; i--){
				if(i<x){
					// console.log("hello y, x and i: "+y+" "+x+" "+i);
					var nextElement = document.getElementById(this.id_gen(y, i));
					if(nextElement!=null){
						left[leftCounter] = nextElement.id;
						if(nextElement.childElementCount!=0){
							break;
						}
						leftCounter++;
					}
				}
			}

			//Going Right
			for(var j=0; j<=7; j++){
				if(j>x){
					var nextElement = document.getElementById(this.id_gen(y, j));
					if(nextElement!=null){
						right[j] = nextElement.id;
						if(nextElement.childElementCount!=0){
							break;
						}
					}
				}
			}

			//Going Down
			for(var k=0; k<=7; k++){
				if(k>y){
					var nextElement = document.getElementById(this.id_gen(k, x));
					if(nextElement!=null){
						down[k] = nextElement.id;
						if(nextElement.childElementCount!=0){
							break;
						}
					}
				}
			}

			//Going Up
			for(var n=7; n>=0; n--){
				if(n<y){
					var nextElement = document.getElementById(this.id_gen(n, x));
					if(nextElement!=null){
						up[upCounter] = nextElement.id;
						if(nextElement.childElementCount!=0){
							break;
						}
						upCounter++;
					}
				}
			}	
		}
		
		left = this.moveArrayToBack(left);
		right = this.moveArrayToBack(right);
		up = this.moveArrayToBack(up);
		down = this.moveArrayToBack(down);

		var horizontal = left.concat(right);
		var vertical = up.concat(down);
		placeIds = horizontal.concat(vertical);
		// console.log("Rook Place Ids: "+placeIds)
	
		return placeIds;
	}

	placeHasCheck(placeId){
		var attackingPlaces = this.checkerGetter.fromPlace(placeId);
		if(attackingPlaces.length>0){
			return true;
		}
		return false;
	}

	rightRookHasMoved(movedPieces){
		for(var i=0; i<movedPieces.length; i++){
			var next = movedPieces[i];
			if(this.isType(next, "player_rook2")){
				return true;
			}
		}
		return false;
	}

	leftRookHasMoved(movedPieces){
		for(var i=0; i<movedPieces.length; i++){
			var next = movedPieces[i];
			if(this.isType(next, "player_rook1")){
				return true;
			}
		}
		return false;
	}

	toRightRookHasPieces(){
		var toRightRookPlaces = ["1F", "1G"];
		var first = document.getElementById(toRightRookPlaces[0]);
		if(first!=null){
			if(first.firstElementChild!=null){
				return true;
			}
		}
		var second = document.getElementById(toRightRookPlaces[1]);
		if(second!=null){
			if(second.firstElementChild!=null){
				return true;
			}
		}
		return false;
	}

	toLeftRookHasPieces(){
		var toLeftRookPlaces = ["1D", "1C", "1B"];
		var first = document.getElementById(toLeftRookPlaces[0]);
		if(first!=null){
			if(first.firstElementChild!=null){
				return true;
			}
		}
		var second = document.getElementById(toLeftRookPlaces[1]);
		if(second!=null){
			if(second.firstElementChild!=null){
				return true;
			}
		}
		var third = document.getElementById(toLeftRookPlaces[2]);
		if(third!=null){
			if(third.firstElementChild!=null){
				return true;
			}
		}
		return false;
	}	

	attackingPlaces(isRook, x, y){
		var attackingRookPlaces = this.getRookMovablePlaces(x, y);
		var rookFwdAttacking = attackingRookPlaces[23];
		var rookBkwdAttacking = attackingRookPlaces[31];
		var rookRightAttacking = attackingRookPlaces[7];
		var rookLeftAttacking = attackingRookPlaces[15];
		var newAttackingRookPlaces = [];
		if(isRook==true){
			this.getAttackingPiecesPlaces(rookFwdAttacking, newAttackingRookPlaces, "comp_rook", "");
			this.getAttackingPiecesPlaces(rookBkwdAttacking, newAttackingRookPlaces, "comp_rook", "");
			this.getAttackingPiecesPlaces(rookRightAttacking, newAttackingRookPlaces, "comp_rook", "");
			this.getAttackingPiecesPlaces(rookLeftAttacking, newAttackingRookPlaces, "comp_rook", "");
		} else {
			this.getAttackingPiecesPlaces(rookFwdAttacking, newAttackingRookPlaces, "", "comp_queen");
			this.getAttackingPiecesPlaces(rookBkwdAttacking, newAttackingRookPlaces, "", "comp_queen");
			this.getAttackingPiecesPlaces(rookRightAttacking, newAttackingRookPlaces, "", "comp_queen");
			this.getAttackingPiecesPlaces(rookLeftAttacking, newAttackingRookPlaces, "", "comp_queen");
		}
		return newAttackingRookPlaces;
	}

	getAttackingPiecesPlaces(placeId, array, type1, type2){
		var type = type1;
		if(type1==""){
			type = type2;
		}
		if(type=="comp_rook"){
		}
		if(placeId!=""){
			var nextPlace = document.getElementById(placeId);
			if(nextPlace.childElementCount!=0){
				var nextPiece = nextPlace.firstElementChild.id;
				if(this.isType(nextPiece, type)){
					array.push(placeId);
				}
			}
		}
	}
}

File: chess_app/ChessEngine/ChessPieces/Rook.py
from ChessPiece import ChessPiece
from chess_app.ChessEngine.Selecter import Select 

class Rook(ChessPiece):
	
	def __init__(self):
		super(Rook, self).__init__()

	def movablePlaces(self, x, y):
		return self.shrinkContinuosArray(self.getRookMovablePlaces(x, y))
	
	# gets all the positions a rook can move gien on coordinates the board 
	def getRookMovablePlaces(self, x, y):
		matrix = self.live_chessboard_matrix
		placeIds = []
		left = ["" for a in range(8)]
		right = ["" for b in range(8)]
		up = ["" for c in range(8)]
		down = ["" for d in range(8)]
		rightCounter = 0
		downCounter = 0

		if x >= 0 and x <= 7 and y >= 0 and y <= 7:

			#Going Left
			for i in range(7, -1, -1):
				if i < x:
					rightSelect = Select()
					nextElement = rightSelect.selectFromParentId(matrix, self.id_gen(y, i))
					if nextElement.parent_id != None:
						right[rightCounter] = nextElement.parent_id
						if nextElement.piece_id != "" and nextElement.piece_id != None:
							break
						rightCounter = rightCounter + 1
					
	

			#Going Right
			for j in range(8):
				if j > x:
					leftSelect = Select()
					nextElement = leftSelect.selectFromParentId(matrix, self.id_gen(y, j))
					if nextElement.parent_id != None:
						left[j] = nextElement.parent_id
						if nextElement.piece_id != "" and nextElement.piece_id != None:
							break
					

			#Going Down
			for k in range(8):
				if k > y:
					upSelect = Select()
					nextElement = upSelect.selectFromParentId(matrix, self.id_gen(k, x))
					if nextElement.parent_id != None:
						up[k] = nextElement.parent_id
						if nextElement.piece_id != "" and nextElement.piece_id != None:
							break
					
					
				
			#Going Up
			for n in range(7, -1, -1):
				if n < y:
					downSelect = Select()
					nextElement = downSelect.selectFromParentId(matrix, self.id_gen(n, x))
					if nextElement.parent_id != None:
						down[downCounter] = nextElement.parent_id
						if nextElement.piece_id != "" and nextElement.piece_id != None:
							break
						downCounter = downCounter + 1
				
					
				
		left = self.moveArrayToBack(left)
		right = self.moveArrayToBack(right)
		up = self.moveArrayToBack(up)
		down = self.moveArrayToBack(down)

		horizontal = left + right
		vertical = up + down
		placeIds = horizontal + vertical
	
		return placeIds
	

	# placeHasCheck(placeId){
	# 	attackingPlaces = self.checkerGetter.fromPlace(placeId)
	# 	if(attackingPlaces.length>0){
	# 		return true
	# 	}
	# 	return false
	# }

	def rightRookHasMoved(self, movedPieces):
		for i in range(len(movedPieces)):
			next_val = movedPieces[i]
			if self.isType(next_val, "player_rook2"):
				return True
			
		return False
	

	def leftRookHasMoved(movedPieces):
		for i in range(len(movedPieces)):
			next_val = movedPieces[i]
			if self.isType(next_val, "player_rook1"):
				return True
		
		return False
	
	# checks if there are pieces to the left of the king
	def toLeftRookHasPieces(self):
		matrix = self.live_chessboard_matrix
		toLeftRookPlaces = ["8F", "8G"]
		
		leftRookSelect1 = Select()
		first = leftRookSelect1.selectFromParentId(matrix, toLeftRookPlaces[0])
		if first.parent_id != None:
			if first.piece_id != None and first.piece_id != "":
				return True
		
		leftRookSelect2 = Select()
		second = leftRookSelect2.selectFromParentId(matrix, toLeftRookPlaces[1])
		if second.parent_id != None:
			if second.piece_id != None and second.piece_id != "":
				return True
			
		return False
	
	# checks if there are pieces to the right of the king
	def toRightRookHasPieces(self):
		matrix = self.live_chessboard_matrix
		toRightRookPlaces = ["8B", "8C", "8D"]
		
		rightRookSelect1 = Select()		
		first = rightRookSelect1.selectFromParentId(matrix, toRightRookPlaces[0])
		if first.parent_id != None:
			if first.piece_id != None and first.piece_id != "":
				return True

		rightRookSelect2 = Select()
		second = rightRookSelect2.selectFromParentId(matrix, toRightRookPlaces[1])
		if second.parent_id != None:
			if second.piece_id != None and second.piece_id != "":
				return True

		rightRookSelect3 = Select()
		third = rightRookSelect3.selectFromParentId(matrix, toRightRookPlaces[2])
		if third.parent_id != None:
			if third.piece_id != None and third.piece_id != "":
				return True

		return False
		
	# sees if a rook or queen is attcking the position of the coordinates given
	def attackingPlaces(self, isRook, x, y):
		attackingRookPlaces = self.getRookMovablePlaces(x, y)
		rookFwdAttacking = attackingRookPlaces[23]
		rookBkwdAttacking = attackingRookPlaces[31]
		rookRightAttacking = attackingRookPlaces[7]
		rookLeftAttacking = attackingRookPlaces[15]
		newAttackingRookPlaces = []
		if isRook == True:
			self.getAttackingPiecesPlaces(rookFwdAttacking, newAttackingRookPlaces, "player_rook", "")
			self.getAttackingPiecesPlaces(rookBkwdAttacking, newAttackingRookPlaces, "player_rook", "")
			self.getAttackingPiecesPlaces(rookRightAttacking, newAttackingRookPlaces, "player_rook", "")
			self.getAttackingPiecesPlaces(rookLeftAttacking, newAttackingRookPlaces, "player_rook", "")
		else:
			self.getAttackingPiecesPlaces(rookFwdAttacking, newAttackingRookPlaces, "", "player_queen")
			self.getAttackingPiecesPlaces(rookBkwdAttacking, newAttackingRookPlaces, "", "player_queen")
			self.getAttackingPiecesPlaces(rookRightAttacking, newAttackingRookPlaces, "", "player_queen")
			self.getAttackingPiecesPlaces(rookLeftAttacking, newAttackingRookPlaces, "", "player_queen")
		
		return newAttackingRookPlaces
	
	# attcking pieces helper
	def getAttackingPiecesPlaces(self, placeId, array, type1, type2):
		matrix = self.live_chessboard_matrix
		piece_type = type1
		if type1 == "":
			piece_type = type2
		
		if placeId != "":
			typeSelect = Select()
			nextPlace = typeSelect.selectFromParentId(matrix, placeId)
			
			if nextPlace.piece_id != None and nextPlace.piece_id != "":
				nextPiece = nextPlace.piece_id
				if self.isType(nextPiece, piece_type):
					array.append(placeId)
				
			
		
	